src/discord/client.rs

```diff
@@ -0,0 +1,167 @@
+//! Discord bot implementations
+use crate::discord::cmd::ping::PingCmd;
+use crate::discord::error::Error as DiscordError;
+use crate::discord::error::ErrorKind::UnknownCommand;
+use crate::discord::metrics::{
+    LABEL_NAME_COMMAND, LABEL_NAME_INTERACTION, LABEL_VALUE_COMMAND_UNKNOWN,
+    METRIC_DISCORD_INTERACTIONS_DURATION, METRIC_DISCORD_INTERACTIONS_TOTAL,
+};
+use crate::discord::utils::interation_name;
+use crate::error::{Error, ErrorKind};
+use metrics::{describe_counter, describe_histogram, histogram, increment_counter, Unit};
+use serenity::model::application::interaction::application_command::ApplicationCommandInteraction;
+use serenity::model::application::interaction::{Interaction, InteractionResponseType};
+use serenity::model::gateway::Ready;
+use serenity::prelude::*;
+use serenity::{async_trait, model::id::GuildId};
+use std::process::exit;
+use std::str::FromStr;
+use std::time::Instant;
+
+use crate::discord::cmd::{CommandExecutable, DiscordCommand};
+use tracing::{debug, error, info, warn};
+
+struct Handler {
+    guild_id: GuildId,
+}
+
+#[async_trait]
+impl EventHandler for Handler {
+    async fn ready(&self, ctx: Context, ready: Ready) {
+        info!("🤝 {} is connected!", ready.user.name);
+
+        let commands = GuildId::set_application_commands(&self.guild_id, &ctx.http, |commands| {
+            commands.create_application_command(|command| {
+                command
+                    .name(DiscordCommand::Ping)
+                    .description("A ping command 🏓 (for testing purposes)")
+            })
+        })
+        .await;
+
+        match commands {
+            Ok(commands) => {
+                info!(
+                    "💻 I now have the following guild slash commands: {:?}",
+                    commands
+                );
+            }
+            Err(err) => {
+                error!("💀 Failed to create application commands: {}", err);
+                exit(-1);
+            }
+        }
+    }
+
+    async fn interaction_create(&self, ctx: Context, interaction: Interaction) {
+        let start = Instant::now();
+
+        let labels = match interaction {
+            Interaction::Ping(_)
+            | Interaction::MessageComponent(_)
+            | Interaction::Autocomplete(_)
+            | Interaction::ModalSubmit(_) => {
+                vec![(LABEL_NAME_INTERACTION, interation_name(&interaction))]
+            }
+            Interaction::ApplicationCommand(ref command) => {
```

Comment:
Look to use the defer resply 
This prevent from Discord telling user that command timedout 
https://docs.rs/serenity/latest/serenity/model/application/interaction/application_command/struct.ApplicationCommandInteraction.html#method.defer

Reply:
Seems that there is a method that control the flag presence